src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractSyntaxTree.java

```diff
@@ -0,0 +1,29 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2022 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.api;
+
+/**
+ * Base interface used for all ASTs of Checkstyle.
+ *
+ * @noinspection MarkerInterface
+ */
+public interface AbstractSyntaxTree {
```

Comment:
I feel that we should have a concrete plan in place with ticket(s) created before we do this.

Reply:
In a perfect world, we should be able to put all methods for getting type, line and column number, iterating through children, etc., in here. But, this would be a significant breaking change.

Reply:
This PR will really not work without this interface as we need to be able to provide the underlying node from the abstracts to the implementations. The only other way would be for the methods to return `Object` and that would require more re-casting in places.

> getting type, line and column number, iterating through children, etc., in here.

This is the plan right now. I mentioned `AstTreeStringPrinter` before and that would be one of the main target areas for this interface. After that, we could simplify both our Walkers (Java and Javadoc) since then they could iterate through different nodes the same way.

All of this future plans really falls under https://github.com/checkstyle/checkstyle/issues/11604  for this interface and what it will do.

> this would be a significant breaking change.

It depends on how we go about this. The current plan is to avoid breaking changes right now by having different method names and deprecating old ways. Eventually, it will lead to a break, but I plan to avoid that as much as I can.

Reply:
> Let's at least explicitly mention this new interface, plans for it, and related details in https://github.com/checkstyle/checkstyle/issues/11604.

I added some of this. Let me know if I am missing anything.

Reply:
Resolved

Reply:
> The only other way would be for the methods to return Object and that would require more re-casting in places.

I was about to suggest this ....
root reason of this update is not very beneficial for Checkstyle itself, BUT we create new contract(interface) with unclear plan. direct users might start to use it and it will complicate their life. BUT this update is not for checkstyle (java) users.

how much extra casts we would need ? 

I would rather "harm" a bit implementation (not a interface), to let some project to reuse us, but not demand any new API for them, as we will not support any not java languages. We can not support even single language reliably, so support of some other is unrealistic. So no extension on api for them, we need our api to work for us first of all. All plans to make api that will work for all AST based validations is not very realistic.

Reply:
can you make one more commit, to show what it will be cost to have Object.

Reply:
> can you make one more commit to show what it will be cost to have Object ?

Done. c692c6835aced11f99d64fe7a77661048d8f4693